### ISSUE TITLE
Update to 74

### ISSUE THREAD

**Issue by Eloston** [ungoogled-software/ungoogled-chromium-portablelinux#5]
Help wanted for updating to ungoogled-chromium 73. Please respond if you'd like to work on this. Thanks

**Comment by emanruse** [comment]
OK. Please share instructions.

**Comment by Eloston** [comment]
@emanruse https://github.com/ungoogled-software/ungoogled-chromium-portablelinux#developer-info

I just updated the `ungoogled-chromium` submodule to 73, so all that should be left is updating patches.

**Comment by emanruse** [comment]
Are these "Developer info" steps which must be completed before or after "Build a tar archive"?

I looked at issue #1 where you discussed with another member and followed the "Build a tar archive" instructions. `build.sh` reaches a point where it asks me which file to patch:

```
INFO: * Applying mojo.patch (6/27)
DEBUG: /usr/bin/patch -p1 --ignore-whitespace -i /tmp/download/ungoogled-chromium-portablelinux/patches/debian_buster/fixes/mojo.patch -d /tmp/download/ungoogled-chromium-portablelinux/build/src --no-backup-if-mismatch --forward
patching file content/shell/BUILD.gn
Hunk #1 FAILED at 125.
Hunk #2 FAILED at 289.
2 out of 2 hunks FAILED -- saving rejects to file content/shell/BUILD.gn.rej
can't find file to patch at input line 27
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|--- a/content/shell/browser/layout_test/layout_test_content_browser_client.cc
|+++ b/content/shell/browser/layout_test/layout_test_content_browser_client.cc
--------------------------
File to patch: 
```

How to continue?

**Comment by Eloston** [comment]
Yes, you have to update the patches before you can attempt a build. "Build a tar archive" is under the section for building.

**Comment by emanruse** [comment]
So I must follow "Updating patches" right before running build.sh?

**Comment by Eloston** [comment]
Well, someone will need to update the patches, otherwise no one can make a build. Once the patches are updated, they won't need to be updated again...

**Comment by emanruse** [comment]
I am not a professional programmer. I thought you just wanted someone to test the process, so here is my result:

https://susepaste.org/view//430b0232

I think "Updating patches" needs fixing. The step `quilt pop -a` must be followed by `cd ../..` in order next steps to work.

Let me know if I can help with something else.

**Comment by nchv** [comment]
@Eloston,

Considering:
- [issue #4](https://github.com/ungoogled-software/ungoogled-chromium-portablelinux/issues/4)
- the current issue
- the fact that on the "main" repo the release version is already 74.0.3729.131-1 and here someone seems fighting to help updating 72 to 73

I am somewhat worried because before this separation of repos it was really easy for me to build latest stable linux_portable ungoogled-chromium. Today it seems not just hard but rather impossible.

I still wonder:
- what made necessary this splitting of repos? (i.e. what good is it considering the above?)
- what the answer to issue #4 is (as having it 2 versions behind seems prone to not being up-to-date with the latest security/bug upstream fixes)

Sorry for the slight off-topic. I hope you can shed some light :)

**Comment by Eloston** [comment]
@nchv 

> what made necessary this splitting of repos? (i.e. what good is it considering the above?)

The overarching reason is it made it easier for other contributors to support their platforms. Under the old monolithic repo, I enforced that patches not be duplicated across all platforms and that packaging scripts conform to an awkward and clunky system I created. Under this new system, Linux distro packagers benefited most because they are free to write packaging scripts and include patches in any manner they want in their own repos (which also makes it easier for other distros with their own Chromium packaging to join in by simply adding a new repo). In the longer term, I am looking to granting push access to some contributors for the platform repos they want to support. I hope this will prevent me from slowing down their development.

This worked well for Linux distros with their own Chromium package, since they can fork the code to support ungoogled-chromium. In addition, platforms with their own set of changes like Windows and macOS remain unaffected by the splitting of repos. However, the separation of development means that each Linux distro must maintain patches separately from others. This doesn't work well for Portable Linux, which depends on a number of patches from other Linux distros.

I don't think there is a solution that can make all platforms happy, but here is what we could do:

1. Manually compare the patches borrowed from the Linux distros and update them
2. Create a repo of patches that can be shared across all or most Linux distros.

**Comment by nchv** [comment]
Thanks for explaining.

Unfortunately I have no idea how to help with anything more than following the instructions which you shared above but another one has already done it. Was that not helpful? I.e. - is there anything else you need in order to move forward to newer version?

**Comment by Eloston** [comment]
@nchv 

> is there anything else you need in order to move forward to newer version?

Nope. I can look into it if I have time, but I cannot make any promises.

**Comment by nchv** [comment]
Thank you!

**Comment by emanruse** [comment]
@Eloston

> However, the separation of development means that each Linux distro must maintain patches separately from others.

Isn't the patching a process applied on the original source code and therefore unrelated to the target OS? In other words - why a patch for (say) Debian should be different from a patch for any other distro (or even for Windows)?

> Create a repo of patches that can be shared across all or most Linux distros.

Then wouldn't it make sense this repo to be prioritized as it is somewhat Linux "universal" (i.e. working for all distros)?

**Comment by Eloston** [comment]
@emanruse 

> Isn't the patching a process applied on the original source code and therefore unrelated to the target OS? In other words - why a patch for (say) Debian should be different from a patch for any other distro (or even for Windows)?

The differences here aren't really in the patches themselves, but more so the development workflow. In the old system, patches that should be shared were explicitly defined as such, or referenced multiple times only in the distros that needed them. As long as the distros require the same kinds of patches over time, then updating the patches for one distro will immediately help other distros using those patches.

But now, the patches are physically separated, so common patches are now duplicated into all distros. So, the maintainers for these distros will have to either update the patch themselves, or copy a new version from another distro. Either way, this is still maintaining the patch separately from others.

> Then wouldn't it make sense this repo to be prioritized as it is somewhat Linux "universal" (i.e. working for all distros)?

Perhaps, but what makes "sense" depends on one's priorities and motivations. Based on my own priorities and motivations, it doesn't make sense for me to work on Portable Linux. It makes more sense if I can help other developers support the platforms they want, because that will encourage more developers to help out.

**Comment by emanruse** [comment]
Why is this (old approach):

> As long as the distros require the same kinds of patches over time, then updating the patches for one distro will immediately help other distros using those patches.

abandoned for (new approach):

> But now, the patches are physically separated, so common patches are now duplicated into all distros. So, the maintainers for these distros will have to either update the patch themselves, or copy a new version from another distro. Either way, this is still maintaining the patch separately from others.

IOW: why is duplication of what is common and working on each duplicate preferred (and supposedly more encouraging for developers)? Isn't that an unnecessary development overhead?

> priorities and motivations

I didn't mean to tell you what to work on :) I was rather wondering how it all works and why. Apart from the "do this and share the result" (what I did here) - is there any documentation explaining in detail:

- what each patch does
- how to update patches properly
- anything else needed for version update

Wouldn't that make it easier to attract more contributors?

**Comment by Eloston** [comment]
> IOW: why is duplication of what is common and working on each duplicate preferred (and supposedly more encouraging for developers)? Isn't that an unnecessary development overhead?

It allows developers to add or update distros at their own pace, in their own way. It doesn't force them to consolidate patches. I've had issues with this in the old system, which is why I refactored everything in the first place.

>  is there any documentation explaining in detail:
> - what each patch does

Nope. The documentation would be high maintenance anyway. Most patches should have comments in them to state what they do.

- how to update patches properly
- anything else needed for version update

All of that is documented in this repo's README. If it's not there, then the documentation should be fixed.

**Comment by emanruse** [comment]
Thanks for explaining!
